executor/index_lookup_join.go

```diff
@@ -0,0 +1,334 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"bytes"
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type keyRowBlock struct {
+	keys        [][]byte
+	rows        []Row
+	requestRows [][]types.Datum
+}
+
+func newKeyRowBlock(batchSize int, hasRequestRow bool) *keyRowBlock {
+	result := &keyRowBlock{
+		keys: make([][]byte, 0, batchSize),
+		rows: make([]Row, 0, batchSize),
+	}
+	if hasRequestRow {
+		result.requestRows = make([][]types.Datum, 0, batchSize)
+	}
+	return result
+}
+
+// Len returns the number of rows.
+func (e *keyRowBlock) Len() int {
+	return len(e.rows)
+}
+
+// Swap implements sort.Interface Swap interface.
+func (e *keyRowBlock) Swap(i, j int) {
+	e.keys[i], e.keys[j] = e.keys[j], e.keys[i]
+	e.rows[i], e.rows[j] = e.rows[j], e.rows[i]
+	if e.requestRows != nil {
+		e.requestRows[i], e.requestRows[j] = e.requestRows[j], e.requestRows[i]
+	}
+}
+
+// Less implements sort.Interface Less interface.
+func (e *keyRowBlock) Less(i, j int) bool {
+	return e.compare(e, i, j) < 0
+}
+
+func (e *keyRowBlock) compare(other *keyRowBlock, i, j int) int {
+	return bytes.Compare(e.keys[i], other.keys[j])
+}
+
+func (e *keyRowBlock) nextBatch(i int) int {
+	for i++; i < len(e.rows); i++ {
+		if e.Less(i-1, i) {
+			return i
+		}
+	}
+	return len(e.rows)
+}
+
+func (e *keyRowBlock) reset() {
+	e.keys = e.keys[:0:cap(e.keys)]
+	e.rows = e.rows[:0:cap(e.rows)]
+	if e.requestRows != nil {
+		e.requestRows = e.requestRows[:0:cap(e.requestRows)]
+	}
+}
+
+// IndexLookUpJoin fetches batches of data from outer executor and constructs ranges for inner executor.
+type IndexLookUpJoin struct {
+	baseExecutor
+
+	outerExec        Executor
+	innerExec        DataReader
+	outerKeys        []*expression.Column
+	innerKeys        []*expression.Column
+	outerFilter      expression.CNFExprs
+	innerFilter      expression.CNFExprs
+	outerOrderedRows *keyRowBlock
```

Comment:
`outerRowBlock` is better? It's not ordered when initialized.

Reply:
I think `outerOrderedRows` is better, it means we should make it ordered before using it.